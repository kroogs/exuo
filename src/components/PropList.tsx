/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 * Copyright Â© 2020 Ty Dira */

import React from 'react'
import { List, ListItem, ListItemText } from '@material-ui/core'
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles'
import { useObserver } from 'mobx-react-lite'
import { getMembers, IAnyStateTreeNode, isStateTreeNode } from 'mobx-state-tree'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {},
    item: {},
    itemText: {},
  }),
)

interface PropListProps {
  model: IAnyStateTreeNode | null
  options?: Partial<{
    groupBy: null
  }>
}

const PropList: React.FunctionComponent<PropListProps> = props => {
  const classes = useStyles()

  return useObserver(() => {
    if (!isStateTreeNode(props.model)) {
      throw Error('Invalid model')
    }

    const members = getMembers(props.model)
    const properties = Object.keys(members.properties)

    return (
      <List className={classes.root}>
        {properties.map(value => (
          <ListItem
            className={classes.item}
            key={value}
            role={undefined}
            dense
            button
          >
            <ListItemText
              className={classes.itemText}
              id={`list-label-${value}`}
              primary={value}
            />
          </ListItem>
        ))}
      </List>
    )
  })
}

export default PropList
