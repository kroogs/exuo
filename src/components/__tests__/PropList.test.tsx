/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 * Copyright Â© 2020 Ty Dira */

import React from 'react'
import { shallow } from 'enzyme'
import { types as t } from 'mobx-state-tree'

import PropList from '../PropList'

test('Renders a list of properties', () => {
  const Model = t.model({
    one: t.number,
    two: t.string,
    three: t.boolean,
  })

  const instance = Model.create({
    one: 100,
    two: 'graphs',
    three: true,
  })

  expect(shallow(<PropList model={instance} />)).toMatchSnapshot()
})
