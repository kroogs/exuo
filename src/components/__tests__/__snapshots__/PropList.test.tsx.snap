// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders a list of properties 1`] = `
<WithStyles(ForwardRef(List))
  className="makeStyles-root-1"
>
  <WithStyles(ForwardRef(ListItem))
    button={true}
    className="makeStyles-item-2"
    dense={true}
    key="one"
  >
    <WithStyles(ForwardRef(ListItemText))
      className="makeStyles-itemText-3"
      id="list-label-one"
      primary="one"
    />
  </WithStyles(ForwardRef(ListItem))>
  <WithStyles(ForwardRef(ListItem))
    button={true}
    className="makeStyles-item-2"
    dense={true}
    key="two"
  >
    <WithStyles(ForwardRef(ListItemText))
      className="makeStyles-itemText-3"
      id="list-label-two"
      primary="two"
    />
  </WithStyles(ForwardRef(ListItem))>
  <WithStyles(ForwardRef(ListItem))
    button={true}
    className="makeStyles-item-2"
    dense={true}
    key="three"
  >
    <WithStyles(ForwardRef(ListItemText))
      className="makeStyles-itemText-3"
      id="list-label-three"
      primary="three"
    />
  </WithStyles(ForwardRef(ListItem))>
</WithStyles(ForwardRef(List))>
`;
